packages/core/core/src/assetUtils.js

```diff
@@ -0,0 +1,180 @@
+export function createAsset(options: AssetOptions): Asset {
+  let idBase = options.idBase != null ? options.idBase : options.filePath;
+  let uniqueKey = options.uniqueKey || '';
+  return {
+    id:
+      options.id != null
+        ? options.id
+        : md5FromString(
+            idBase + options.type + getEnvironmentHash(options.env) + uniqueKey,
+          ),
+    committed: options.committed ?? false,
+    hash: options.hash,
+    filePath: options.filePath,
+    isIsolated: options.isIsolated == null ? false : options.isIsolated,
+    isInline: options.isInline == null ? false : options.isInline,
+    isSplittable: options.isSplittable,
+    type: options.type,
+    contentKey: options.contentKey,
+    mapKey: options.mapKey,
+    astKey: options.astKey,
+    astGenerator: options.astGenerator,
+    dependencies: options.dependencies || new Map(),
+    includedFiles: options.includedFiles || new Map(),
+    isSource: options.isSource,
+    outputHash: options.outputHash,
+    pipeline: options.pipeline,
+    env: options.env,
+    meta: options.meta || {},
+    stats: options.stats,
+    symbols: options.symbols || new Map(),
+    sideEffects: options.sideEffects != null ? options.sideEffects : true,
+    uniqueKey: uniqueKey,
+    plugin: options.plugin,
+    configPath: options.configPath,
+  };
+}
+
+const generateResults: WeakMap<Asset, Promise<GenerateOutput>> = new WeakMap();
+
+export function generateFromAST(
+  asset: CommittedAsset | UncommittedAsset,
+): Promise<GenerateOutput> {
+  let output = generateResults.get(asset.value);
+  if (output == null) {
+    output = _generateFromAST(asset);
+    generateResults.set(asset.value, output);
+  }
+  return output;
+}
+
+async function _generateFromAST(asset: CommittedAsset | UncommittedAsset) {
+  let ast = await asset.getAST();
+  if (ast == null) {
+    throw new Error('Asset has no AST');
+  }
+
+  let pluginName = nullthrows(asset.value.plugin);
+  let plugin: Transformer = await loadPlugin(
+    asset.options.packageManager,
+    pluginName,
+    nullthrows(asset.value.configPath),
+  );
+  if (!plugin.generate) {
+    throw new Error(`${pluginName} does not have a generate method`);
+  }
+
+  let {code, map} = await plugin.generate({
+    asset: new PublicAsset(asset),
+    ast,
+    options: new PluginOptions(asset.options),
+    logger: new PluginLogger({origin: pluginName}),
+  });
+
+  let mapBuffer = map?.toBuffer();
+  // Store the results in the cache so we can avoid generating again next time
+  await Promise.all([
+    asset.options.cache.setStream(
+      nullthrows(asset.value.contentKey),
+      blobToStream(code),
+    ),
+    mapBuffer == null
+      ? Promise.resolve()
+      : asset.options.cache.setBlob(nullthrows(asset.value.mapKey), mapBuffer),
+  ]);
+
+  return {
+    code:
+      code instanceof Readable
+        ? asset.options.cache.getStream(nullthrows(asset.value.contentKey))
+        : code,
+    map,
+  };
+}
+
+export async function getConfig(
```

Comment:
We should only expose this in transformers IMO

Reply:
`getPackage` is probably useful everywhere though.

Reply:
Validators receive an `Asset` (not a `MutableAsset` unlike transformers), and use `.getConfig()` 😕